src/gz21_ocean_momentum/lib/data.py

```diff
@@ -0,0 +1,293 @@
+    #                      yt_ocean=u_v_dataset.coords['yu_ocean'])
+    # u_v_dataset['temp'] = u_v_dataset['surface_temperature'].interp(
+    #     interp_coords)
+
+    # High res advection terms
+    adv = _advections(u_v_dataset, grid_data)
+    # Filtered advections
+    filtered_adv = _spatial_filter_dataset(adv, grid_data, float(scale)/2)
+
+    # Filtered u,v field and temperature
+    u_v_filtered = _spatial_filter_dataset(u_v_dataset, grid_data, float(scale)/2)
+    # Advection term from filtered velocity field
+    adv_filtered = _advections(u_v_filtered, grid_data)
+
+    # Forcing
+    ds_forcing = adv_filtered - filtered_adv
+    ds_forcing = ds_forcing.rename({"adv_x": "S_x", "adv_y": "S_y"})
+    # Merge filtered u,v, temperature and forcing terms
+    ds_merged = ds_forcing.merge(u_v_filtered)
+    logger.debug("uncoarsened forcings follow below:")
+    logger.debug(ds_merged)
+
+    # Coarsen
+    ds_merged_coarse = ds_merged.coarsen(
+        {"xu_ocean": scale, "yu_ocean": scale}, boundary="trim"
+    ).mean()
+
+    if nan_or_zero == "zero":
+        # Replace zeros with nans for consistency
+        ds_merged_coarse = ds_merged_coarse.where(ds_merged_coarse["usurf"] != 0)
+
+    # Specify input vs output type for each variable of the dataset. Might
+    # be used later on for training or testing.
+    ds_merged_coarse["S_x"].attrs["type"] = "output"
+    ds_merged_coarse["S_y"].attrs["type"] = "output"
+    ds_merged_coarse["usurf"].attrs["type"] = "input"
+    ds_merged_coarse["vsurf"].attrs["type"] = "input"
+
+    return ds_merged_coarse
+
+
+def _advections(u_v_field: xr.Dataset, grid_data: xr.Dataset) -> xr.Dataset:
+    """
+    Compute advection terms corresponding to the passed velocity field.
+
+    Parameters
+    ----------
+    u_v_field : xarray Dataset
+        Velocity field, must contains variables "usurf" and "vsurf", coordinates
+        "xu_ocean" and "yu_ocean"
+    grid_data : xarray Dataset
+        grid data, must contain variables "dxu" and "dyu"
+
+    Returns
+    -------
+    result : xarray Dataset
+        Advection components, under variable names "adv_x" and "adv_y"
+    """
+    dxu = grid_data["dxu"]
+    dyu = grid_data["dyu"]
+    gradient_x = u_v_field.diff(dim="xu_ocean") / dxu
+    gradient_y = u_v_field.diff(dim="yu_ocean") / dyu
+    # Interpolate back the gradients
+    interp_coords = {
+        "xu_ocean": u_v_field.coords["xu_ocean"],
+        "yu_ocean": u_v_field.coords["yu_ocean"],
+    }
+    gradient_x = gradient_x.interp(interp_coords)
+    gradient_y = gradient_y.interp(interp_coords)
+    u, v = u_v_field["usurf"], u_v_field["vsurf"]
+    adv_x = u * gradient_x["usurf"] + v * gradient_y["usurf"]
+    adv_y = u * gradient_x["vsurf"] + v * gradient_y["vsurf"]
+    result = xr.Dataset({"adv_x": adv_x, "adv_y": adv_y})
+    # check if we can simply prevent the previous operation from adding chunks
+    # result = result.chunk(dict(xu_ocean=-1, yu_ocean=-1))
```

Comment:
I don't know what rechunking to size -1 does. I don't think these operations should add a chunk, we're not working on the chunked dimension (time). Even less likely with moving back to `xarray.map_blocks`: this is an inner function, so it will similarly be called once per chunk(=time point).

So I'm happy to remove it the commented-out code. We could leave a better comment in place, like `# can these ops add chunks to our underlying Dask array?`

Reply:
Yes, better comment sounds good